lib/plugin.js

```diff
@@ -0,0 +1,142 @@
+'use strict';
+
+const {createUnsupportedError} = require('./errors');
+const {defineConstants, castArray} = require('./utils');
+
+const constants = (exports.constants = defineConstants({
+  PLUGIN_ROOT_HOOKS: 'mochaHooks',
+  PLUGIN_GLOBAL_SETUP: 'mochaGlobalSetup',
+  PLUGIN_GLOBAL_TEARDOWN: 'mochaGlobalTeardown'
+}));
+const {
+  PLUGIN_ROOT_HOOKS,
+  PLUGIN_GLOBAL_SETUP,
+  PLUGIN_GLOBAL_TEARDOWN
+} = constants;
+const PLUGIN_NAMES = new Set(Object.values(constants));
+
+exports.PluginLoader = class PluginLoader {
+  constructor({
+    pluginNames = PLUGIN_NAMES,
+    pluginValidators = PluginValidators
+  } = {}) {
+    this.pluginNames = Array.from(pluginNames);
+    this.pluginMap = new Map(
+      this.pluginNames.map(pluginName => [pluginName, []])
+    );
+    this.pluginValidators = pluginValidators;
+  }
+
+  load(requiredModule) {
+    // we should explicitly NOT fail if other stuff is exported.
+    // we only care about the plugins we know about.
+    if (requiredModule && typeof requiredModule === 'object') {
+      return this.pluginNames.reduce((isFound, pluginName) => {
+        const plugin = requiredModule[pluginName];
+        if (plugin) {
+          if (this.pluginValidators[pluginName]) {
+            this.pluginValidators[pluginName](plugin);
+          }
+          this.pluginMap.set(pluginName, [
+            ...this.pluginMap.get(pluginName),
+            ...castArray(plugin)
+          ]);
+          return true;
+        }
+        return isFound;
+      }, false);
+    }
+    return false;
+  }
+
+  async finalize() {
+    const mochaHooks = this.pluginMap.get(PLUGIN_ROOT_HOOKS);
```

Comment:
I should add this in a code comment, but the idea here is that we call `load()` on all required modules, which fishes the special named exports out of those modules, and then calling `finalize()` aggregates/flattens everything into a single data structure, which is suitable for passing into `Mocha` constructor